dali/kernels/test/warp_test/warp_gpu_test.cu

```diff
@@ -0,0 +1,307 @@
+  ASSERT_EQ(cpu_out.shape[1], img_tensor.shape[0]);
+  ASSERT_EQ(cpu_out.shape[2], 3);
+
+  for (int y = 0; y < cpu_out.shape[0]; y++) {
+    for (int x = 0; x < cpu_out.shape[1]; x++) {
+      for (int c = 0; c < 3; c++) {
+        EXPECT_EQ(*cpu_out(y, x, c), *cpu_img(x, y, c));
+      }
+    }
+  }
+}
+
+TEST(WarpGPU, Affine_Transpose_Single) {
+  AffineMapping2D mapping_cpu = mat2x3{{
+    { 0, 1, 0 },
+    { 1, 0, 0 }
+  }};
+
+  cv::Mat cv_img = cv::imread(testing::dali_extra_path() + "/db/imgproc/alley.png");
+  auto cpu_img = view_as_tensor<uint8_t>(cv_img);
+  auto gpu_img = copy<AllocType::GPU>(cpu_img);
+  auto img_tensor = gpu_img.first;
+
+  TensorListView<StorageGPU, uint8_t, 3> in_list;
+  in_list.resize(1, 3);
+  in_list.shape.set_tensor_shape(0, img_tensor.shape);
+  in_list.data[0] = img_tensor.data;
+
+  WarpGPU<AffineMapping2D, 2, uint8_t, uint8_t, BorderClamp, DALI_INTERP_NN> warp;
+
+  ScratchpadAllocator scratch_alloc;
+
+  auto mapping_gpu = memory::alloc_unique<AffineMapping2D>(AllocType::GPU, 1);
+  TensorShape<2> out_shape = { img_tensor.shape[1], img_tensor.shape[0] };
+  KernelContext ctx = {};
+  auto out_shapes_hw = make_span<1>(&out_shape);
+  auto mappings = make_tensor_gpu<1>(mapping_gpu.get(), { 1 });
+  copy(mappings, make_tensor_cpu<1>(&mapping_cpu, { 1 }));
+
+  KernelRequirements req = warp.Setup(ctx, in_list, mappings, out_shapes_hw);
+  scratch_alloc.Reserve(req.scratch_sizes);
+  TestTensorList<uint8_t, 3> out;
+  out.reshape(req.output_shapes[0].to_static<3>());
+  auto scratchpad = scratch_alloc.GetScratchpad();
+  ctx.scratchpad = &scratchpad;
+  warp.Run(ctx, out.gpu(0), in_list, mappings, out_shapes_hw);
+
+  auto cpu_out = out.cpu(0)[0];
+  cudaDeviceSynchronize();
+  ASSERT_EQ(cpu_out.shape[0], img_tensor.shape[1]);
+  ASSERT_EQ(cpu_out.shape[1], img_tensor.shape[0]);
+  ASSERT_EQ(cpu_out.shape[2], 3);
+
+  int errors = 0;
+  int printed = 0;
+  for (int y = 0; y < cpu_out.shape[0]; y++) {
+    for (int x = 0; x < cpu_out.shape[1]; x++) {
+      for (int c = 0; c < 3; c++) {
+        if (*cpu_out(y, x, c) != *cpu_img(x, y, c)) {
+          if (errors++ < 100) {
+            printed++;
+            EXPECT_EQ(*cpu_out(y, x, c), *cpu_img(x, y, c))
+              << "@ x = " << x << " y = " << y << " c = " << c;
+          }
+        }
+      }
+    }
+  }
+  if (printed != errors) {
+    FAIL() << (errors - printed) << " more erors.";
+  }
+}
+
+/// @brief Apply correction of pixel centers and convert the mapping to
+///        OpenCV matrix type.
+inline cv::Matx<float, 2, 3> AffineToCV(const AffineMapping2D &mapping) {
+  vec2 translation = mapping({0.5f, 0.5f}) - vec2(0.5f, 0.5f);
+  mat2x3 tmp = mapping.transform;
+  tmp.set_col(2, translation);
+
+  cv::Matx<float, 2, 3> cv_transform;
+  for (int i = 0; i < 2; i++)
+    for (int j = 0; j < 3; j++)
+      cv_transform(i, j) = tmp(i, j);
+  return cv_transform;
+}
+
+inline void DumpImages(const std::string &base_name,
```

Comment:
We have some utilities functions in dali/util/image.h. You can use one of it or move this one there.

Reply:
It's rather a debugging utility. Also, the interface dependens on `cv::Mat`, which I wouldn't like to have in our util/image header. I may move it to some test utils, though.